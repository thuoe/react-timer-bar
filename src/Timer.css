@import url('https://fonts.googleapis.com/css2?family=Roboto:wght@100;400&display=swap');

.timer-container {
  display: block;
  background-color: red;
  border-radius: 3px;
  width: 300px;
  height: 100px;
  text-align: center;
}

.timer-bar { 
  position: relative;
  height: 100%;
  width: 0;
  border-radius: 3px;
  transition: width 200ms;
} 

span {
  font-family: 'Roboto', sans-serif;
  text-align: center;
}
