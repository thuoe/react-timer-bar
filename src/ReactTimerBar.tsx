import * as React from 'react';
import './Timer.css';

type ReactTimerProps = {
  time: number;
  fillColor: string;
};

export const ReactTimerBar = ({ time = 0, fillColor = 'blue' }: ReactTimerProps): JSX.Element => {
  const timerBar = React.useRef<HTMLDivElement | null>(null);
  const [timeElapsed, setTimeElapsed] = React.useState<number>(0);

  function formatTime(newTime: number): string {
    const minutes = Math.floor(newTime / 60);
    const seconds = newTime - minutes * 60; 
    const secondPrefix = seconds < 10 ? '0' : '';   
    return `${minutes}:${secondPrefix}${seconds}`;
  }

  React.useEffect(() => {
    setTimeElapsed(0);
  }, [time]);

  React.useEffect(() => {
    let timerInterval: NodeJS.Timer;
    if (timeElapsed < time) {
      timerInterval = setInterval(() => {
        setTimeElapsed(timeElapsed + 1);
      }, 1000);
    }
    return () => {
      if (timerInterval) {
        clearInterval(timerInterval);
      }
    };
  }, [timeElapsed]);

  const timerWidth = timerBar.current ? timerBar.current.offsetWidth : 0;
  const fillWidth = (timeElapsed / time) * timerWidth;

  return (
    <>
      <div ref={timerBar} className="timer-container">
        <div className="timer-bar" style={{ backgroundColor: fillColor, width: `${fillWidth}px` } }></div>
        <span title="Time Elapsed">Time Elapsed {formatTime(time)}</span>
      </div>
    </>
  ); 
};