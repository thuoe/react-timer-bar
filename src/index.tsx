import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ReactTimerBar } from './ReactTimerBar';

ReactDOM.render(<ReactTimerBar fillColor="blue" time={10} />, document.getElementById('output'));
