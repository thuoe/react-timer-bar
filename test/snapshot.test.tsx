import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { ReactTimerBar } from '../src';


describe('Basic snapshot test', () => {

  it('renders succesfully', () => {
    const tree = renderer
      .create(<ReactTimerBar time={30} fillColor='red' />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});
