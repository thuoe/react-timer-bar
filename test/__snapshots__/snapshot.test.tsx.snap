// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic snapshot test renders succesfully 1`] = `
<div
  className="css-1991io3"
>
  <div
    className="css-1km8or2"
  />
  <span
    title="Time Elapsed"
  >
    Time Elapsed 
    0:30
  </span>
</div>
`;
